libcli.c

```diff
@@ -1966,11 +1990,1146 @@ void cli_set_idle_timeout_callback(struct cli_def *cli, unsigned int seconds, in
+
+    if (!strcmp(c->name, name)) {
+      *p = c->next;
+      free_z(c->name);
+      free_z(c->value);
+      free_z(c);
+    } else {
+      p = &(*p)->next;
+    }
+  }
+}
+
+int cli_int_add_optarg_value(struct cli_def *cli, const char *name, const char *value, int allow_multiple) {
+  struct cli_optarg_pair *optarg_pair, **anchor;
+  int rc = CLI_ERROR;
+
+  for (optarg_pair = cli->found_optargs, anchor = &cli->found_optargs; optarg_pair;
+       anchor = &optarg_pair->next, optarg_pair = optarg_pair->next) {
+    // break if we found this name *and* allow_multiple is false
+    if (!strcmp(optarg_pair->name, name) && !allow_multiple) {
+      break;
+    }
+  }
+  // if we *didn't* find this, then allocate a new entry before proceeding
+  if (!optarg_pair) {
+    optarg_pair = (struct cli_optarg_pair *)calloc(1, sizeof(struct cli_optarg_pair));
+    *anchor = optarg_pair;
+  }
+  // set the value
+  if (optarg_pair) {
+    // name is null only if we didn't find it
+    if (!optarg_pair->name) optarg_pair->name = strdup(name);
+
+    // value may be overwritten, so free any old value.
+    if (optarg_pair->value) free_z(optarg_pair->value);
+    optarg_pair->value = strdup(value);
+
+    rc = CLI_OK;
+  }
+  return rc;
+}
+
+struct cli_optarg_pair *cli_get_all_found_optargs(struct cli_def *cli) {
+  if (cli) return cli->found_optargs;
+  return NULL;
+}
+
+char *cli_get_optarg_value(struct cli_def *cli, const char *name, char *find_after) {
+  char *value = NULL;
+  struct cli_optarg_pair *optarg_pair;
+
+  printf("cli_get_optarg_value entry - looking for <%s> after <%p>\n", name, (void *)find_after);
+  for (optarg_pair = cli->found_optargs; !value && optarg_pair; optarg_pair = optarg_pair->next) {
+    printf("  Checking %s with value %s <%p> \n", optarg_pair->name, optarg_pair->value, (void *)optarg_pair->value);
+
+    // check next entry if this isn't our name
+    if (strcasecmp(optarg_pair->name, name)) continue;
+
+    // did we have a find_after, then ignore anything up until our find_after match
+    if ((find_after) && (optarg_pair->value == find_after)) {
+      find_after = NULL;
+      continue;
+    } else if (!find_after) {
+      value = optarg_pair->value;
+    }
+  }
+  printf("cli_get_optarg_value exit - returning <%s><%p>\n", name, value);
+  return value;
+}
+
+void cli_int_free_buildmode(struct cli_def *cli) {
+
+  if (!cli || !cli->buildmode) return;
+  //  cli_unregister_tree(cli, cli->commands, CLI_BUILDMODE_COMMAND);
+  cli->mode = cli->buildmode->mode;
+  free_z(cli->buildmode->cname);
+  free_z(cli->buildmode->mode_text);
+  cli_int_free_found_optargs(&cli->buildmode->found_optargs);
+  free_z(cli->buildmode);
+}
+
+int cli_int_enter_buildmode(struct cli_def *cli, struct cli_pipeline_stage *stage, char *mode_text) {
+  struct cli_optarg *optarg;
+  struct cli_command *c;
+  struct cli_buildmode *buildmode;
+
+  if ((!cli) || !(buildmode = (struct cli_buildmode *)calloc(1, sizeof(struct cli_buildmode)))) {
```

Comment:
Remove unnecessary parens around `(!cli)`

Reply:
Fixed.  The (!cli) had another clause in there at one point.  Just didn't rip the parens off.